### PR TITLE
Music: new instrument grid editor

### PR DESCRIPTION
**Pull request by sanchitmalhotra126**
A proposal for a new instrument grid editor, meant to be used for the play tune and play drums blocks. This was influenced by [ongoing discussions](https://docs.google.com/document/d/185m-gy8h2TaC_7z1nbFM82FGS11Onv_E30Q8uoADBag/edit?tab=t.0) around the play tune block, investigating a few directions we've been wanting to explore. Key callouts:

- Two editor types, **notes** and **drums**
- **Notes Editor**
  - Has two modes, "All Notes" and "Best Notes"
  - "Best Notes" displays only notes that work with the key of the selected song, while "All Notes" displays all notes
  - "Best Notes" uses a cool to warm color palette for low to high notes

- **Drums Editor**
  - Displays drum notes in reverse order, matching the notes editor and most modern drum editors (bass drum at the bottom, cymbals at the top).
  - Uses a cool tone color palette

**Code Factoring**
Because both the play tune and play drums block [now use a common interface](https://github.com/code-dot-org/code-dot-org/pull/61602) (InstrumentEvent), we can technically deprecate the separate TunePanel and PatternPanel components, in favor of rendering the InstrumentGrid component directly. To avoid updating existing components, I only removed TunePanel and kept PatternPanel as-is. With the `?play-tune-block=true` URL flag, the pattern panel (play drums block) **will also use the new InstrumentGrid**. I'm open to creating a separate URL flag to control this but thought it would be simpler to keep it tied to the same flag for now. Without the flag, users should see the existing pattern panel UI, so as to not affect any active production users.

**Open Questions**
There are a few areas that are still unresolved that I'm hoping we can solve through playtesting and iteration:
- What should the default notes mode be?
- What happens to your pattern value when you switch between All Notes/Best Notes? Do we keep the non-key notes but not render them? Should we save extra state to know which mode you were in?
- Where should we scroll to when opening the editor? Currently we don't do any scroll so you start at the higher end of the note range. Should we compute and scroll to the middle octave?

And of course, all of the UI, text, colors, etc are all very much up for review here :) I'm hoping that this can serve as a jumping off point for further discussion, but don't expect any of this be final.

<details>
<summary>Screenshots</summary>

Screenshot 2025-02-11 at 4 07 13 PM
Screenshot 2025-02-11 at 4 07 23 PM
Screenshot 2025-02-11 at 4 07 31 PM
</details>

<details>
<summary>Videos</summary>

https://github.com/user-attachments/assets/e0eabe43-5436-411c-8f35-a606c9ec4dcb

https://github.com/user-attachments/assets/20fdce00-02e7-4e8d-8f96-40d48de8d445

https://github.com/user-attachments/assets/b6a71301-6b33-41f8-9868-dd4b9a1b0bda
</details>

<details>
<summary>Old Description (OUTDATED)</summary>

A prototype exploring a couple new directions for our instrument editor UI, notably the play tune block. This started as an attempt to make a "simple" version of the tune block and slowly started to accumulate a few other features as well. The main changes include:
- The editor defaults to a "simple" mode that:
  - is limited to a single octave
  - only includes notes that work with the key of the song
  - uses a cool to warm color palette for low to high notes
  - uses 8th notes instead of 16th notes to reduce the number of cells on screen
  - [[image (outdated)](https://github.com/user-attachments/assets/5706c03e-dad3-4dfb-a3f3-a218a35c9bab)]

- Additionally, you can toggle to a "chromatic" mode that allows full chromatic, 16th note editing. Only one octave is displayed on screen at a time, but you can use the page/up down buttons on the side to switch octaves. This technically works in simple mode too.
  - [[image (outdated)](https://github.com/user-attachments/assets/fc428f11-55d4-4a67-b016-d4aa29814671)]

- I also added a drum mode which is similar to the chromatic mode but with only the set of available drum notes and a different color palette. As an experiment in this PR, I've replaced the pattern panel UI with this drum editor UI.
  - [[image (outdated)](https://github.com/user-attachments/assets/efdaaf0c-6888-42af-9d27-ee83b7fbe947)]

One additional change is that I added a small amount of reverb to the default instrument output to give it a little bit of pop. I find that this makes the instruments sound a lot better and feels a lot better to edit and preview.

Demo video showcasing interactions and previewing (no sound): [[video (outdated)](https://github.com/user-attachments/assets/296f98d8-42e0-4b99-8d7c-2270dcf8dc0c)]

The code is still a bit rough and would need to be cleaned up before merging, and of course the UI is not meant to be final, but to showcase some ideas.

</details>

### CODE REVIEW

File: dashboard/app/assets/stylesheets/application.scss [hunk 1]
```diff
@@ -2884,6 +2884,14 @@ a.download-video {
   stroke: none;
 }
 
+// Allows certain custom Blockly fields to use the full height of their
+// content, if the specific full height class name is applied.
+.blocklyDropDownContent {
```
Comment (sanchitmalhotra126) [inline dashboard/app/assets/stylesheets/application.scss]: Will pull this out into a separate change

Edit: actually, the editor isn't very useable without this change, so I'll keep it in here. It also demonstrates how this class is meant to be used.

File: apps/src/music/player/ToneJSPlayer.ts [hunk 2]
```diff
@@ -127,10 +128,18 @@ class ToneJSPlayer {
       }
     });
 
+    // TODO: Put behind URL flag
```
Comment (sanchitmalhotra126) [inline apps/src/music/player/ToneJSPlayer.ts]: Will pull this out into a separate change

Comment (sanchitmalhotra126) [reply apps/src/music/player/ToneJSPlayer.ts]: Reverted. Will create a separate PR for this.

File: apps/src/music/views/PreviewControlsV2.tsx
Comment (sanchitmalhotra126) [inline apps/src/music/views/PreviewControlsV2.tsx:1]: The hope is to replace the existing `PreviewControls` component with this since it uses DSCO buttons (and is overall smaller), but keeping it separate right now to not affect live production UI.

**Comment by breville** [comment]
This is great!  Here are some notes that I took while using it:

- The AI drums grid has special layout so that there aren't non-clickable gaps between columns in the same measure or between rows.  This cut down on annoying clicks-that-do-nothing.  I think it would be worth using here too.
- At least for the default pack's key, notes in the very top row of "Best Notes" aren't visible in "All Notes".
- I would recommend a starting scroll position in the middle, with care to have a "peek" (i.e. partial) row for the top-most and bottom-most visible rows.  It could also be nice to play a scroll animation to ease to here, especially when first opening (i.e. when there are not yet any notes placed).  
- The dark background seems to produce a dark scroll bar on Safari that isn't easy to see.  This can be remedied with a `color-scheme: dark` CSS attribute (TIL) or possibly with a background treatment similar to that in other panels.
- In the "All Notes" view, the hover color overrides the selected color.
- The light color of the disabled play icon feels a bit distracting.  Maybe there is a dark mode variation?  But given that we don't disable the clear button, maybe we could leave both buttons enabled?  Certainly it's still informative to play nothing when first exploring the interface.
- Selecting A#5 in "All Notes" shows nothing in "Best Notes", but still plays.  It seems to make sense that it shouldn't play if you switch to "Best Notes".  Ideally it would still be there if you switch back to "All Notes" so that you don't lose a lot of work due to a quick toggle back and forth.
- It feels like the panel should stay the same size when toggling between modes.
- The expanded size seems too big for many of our devices, which are around 1366x605.  The AI drums panel is very shallow for this reason.
- Reversing the drums makes sense to match industry patterns, but I presume most students using this block will start from the top down, and many of them will have learnt drums from the AI drums block last year.  For those reasons, should we keep the existing top-down order?

**Comment by sanchitmalhotra126** [comment]
@breville thanks a bunch for the detailed feedback!

> * The AI drums grid has special layout so that there aren't non-clickable gaps between columns in the same measure or between rows.  This cut down on annoying clicks-that-do-nothing.  I think it would be worth using here too.

Oh interesting, curious about this - do you mean the gaps between cells? These shouldn't be coming up as clickable in theory, but wonder if it's something browser specific...?

> * At least for the default pack's key, notes in the very top row of "Best Notes" aren't visible in "All Notes".

Great catch. Will adjust this!

> * I would recommend a starting scroll position in the middle, with care to have a "peek" (i.e. partial) row for the top-most and bottom-most visible rows.  It could also be nice to play a scroll animation to ease to here, especially when first opening (i.e. when there are not yet any notes placed).

Agreed, starting at the middle octave seems like the right move. I'll look into this.

> * The dark background seems to produce a dark scroll bar on Safari that isn't easy to see.  This can be remedied with a `color-scheme: dark` CSS attribute (TIL) or possibly with a background treatment similar to that in other panels.

Good call, will try the `color-scheme` method! Also yeah - worth calling out that that the existing play notes and tunes panel has a slightly lighter grey background, while the existing drums panel has this darker (neutral_dark) background. I'll change to the lighter one and see how it appears.

> * In the "All Notes" view, the hover color overrides the selected color.

Good catch, fill fix.

> * The light color of the disabled play icon feels a bit distracting.  Maybe there is a dark mode variation?  But given that we don't disable the clear button, maybe we could leave both buttons enabled?  Certainly it's still informative to play nothing when first exploring the interface.

Good call. Will play around with some different button colors.

> * Selecting A#5 in "All Notes" shows nothing in "Best Notes", but still plays.  It seems to make sense that it shouldn't play if you switch to "Best Notes".  Ideally it would still be there if you switch back to "All Notes" so that you don't lose a lot of work due to a quick toggle back and forth.

Yeah, we'll need to think through a strategy on how to best handle switching between modes. We might need to start saving the selection type to the interface. Maybe we can address this holistically as a follow-up?

> * It feels like the panel should stay the same size when toggling between modes.
> * The expanded size seems too big for many of our devices, which are around 1366x605.  The AI drums panel is very shallow for this reason.

Good points. Will play around with size.

> * Reversing the drums makes sense to match industry patterns, but I presume most students using this block will start from the top down, and many of them will have learnt drums from the AI drums block last year.  For those reasons, should we keep the existing top-down order?

Fair enough. Not a strong preference here; it definitely feels more natural for me to reverse the order, but acknowledging that it's likely due to familiarity with other tools. Happy to reverse or keep as is!

**Comment by sanchitmalhotra126** [comment]
Updates:

- Fixed the highest note mismatch between All and Best notes
- Editor now smooth scrolls to the middle octave when first opened. I include a little peek amount above and below to hopefully convey that there are more notes
- Updated background color, and used `color-scheme: dark` for dark scrollbars
- Fixed "All notes" hover color
- Moved Preview Controls to top section and changed button colors
- Overall reduced size, tightened padding and reduced cell height a tad. I also made the width of the All Notes and Best Note labels the same to reduce jumping when switching between modes.

Screenshots (browser window is 1366x605)
Screenshot 2025-02-12 at 12 12 46 PM
Screenshot 2025-02-12 at 12 12 58 PM

**Comment by dju90** [comment]
Can we call "best notes" "simple notes" ?

**Comment by sanchitmalhotra126** [comment]
> Can we call "best notes" "simple notes" ?

@dju90 curious to discuss this; I agree that "Best Notes" likely isn't the right name, but also worried that "Simple Notes" may come off as a bit pejorative? Wouldn't want users to feel that the "Simple" mode is limited in some way, or that they need to graduate to the more "advanced" mode...

Something like "Notes in Key" / "Notes in Song"? Though that also feels a bit wordy

**Comment by sanchitmalhotra126** [comment]
@breville @dju90 latest update includes:

- Fix for Brendan's comment:
  > The AI drums grid has special layout so that there aren't non-clickable gaps between columns in the same measure or between rows. This cut down on annoying clicks-that-do-nothing. I think it would be worth using here too.
- Even smaller panel height. Particularly curious about this because this brings us right under Blockly's default 300px ,ax height, which currently I'm overriding (if we like this, I can remove the extra code needed to override the max height value).

Screenshot 2025-02-12 at 2 45 19 PM

**Comment by sanchitmalhotra126** [comment]
The latest changes include adopting @breville's `EaseIntoView` component to perform a smooth scroll from the bottom to the middle octave. I made a few small tweaks to the component to support stopping at a specific position rather than at the top.

Also reverted the Blockly max height related changes as the panel is now under Blockly's editor max height of 300px.

**Review by breville — APPROVED** [review]
This is great!

**Review by dju90 — COMMENTED** [review]
double-scroll-play-tune
I know this is going behind a flag, but imo I'm now encountering a double scroll bar experience (one for the grid and one for the entire panel) and I don't think that's ideal.
...though maybe I just have'nt pulled the latest and need to do that first.

**Comment by sanchitmalhotra126** [comment]
> though maybe I just have'nt pulled the latest and need to do that first.

Hmm yeah, the latest changes should have the preview buttons at the top and the overall size should be reduced, so the outer scrollbar shouldn't be appearing. Let me know if it still happens after you pull though!
